Metric spaces (#1162)
UniMath/agda-unimath: src/metric-spaces/limits-of-cauchy-approximations-in-premetric-spaces.lagda.md

open import foundation.dependent-pair-types
open import foundation.equivalences
open import foundation.function-extensionality
open import foundation.function-types
open import foundation.homotopies
open import foundation.identity-types
open import foundation.logical-equivalences
open import foundation.propositions
open import foundation.subtypes
open import foundation.transport-along-identifications
open import foundation.universe-levels

open import metric-spaces.cauchy-approximations-premetric-spaces
open import metric-spaces.extensional-premetric-structures
open import metric-spaces.premetric-spaces
open import metric-spaces.short-functions-premetric-spaces
open import metric-spaces.symmetric-premetric-structures
open import metric-spaces.triangular-premetric-structures
```

</details>

## Idea

In a [premetric space](metric-spaces.premetric-spaces.md) `A`, an element `l` of
the carrier type of `A` is a
{{#concept "limit" Disambiguation="of a Cauchy approximation in a premetric space" Agda=is-limit-cauchy-approximation-Premetric-Space}}
of a
[Cauchy approximation](metric-spaces.cauchy-approximations-premetric-spaces.md)
`f` in `A` if `f ε` is an

> [`(ε + δ)`-neighbor](metric-spaces.premetric-structures.md) of `l` for all

Reviewer: ```suggestion
`(ε + δ)`-[neighbor](metric-spaces.premetric-structures.md) of `l` for all
```

[positive rationals](elementary-number-theory.positive-rational-numbers.md)

> `(ε δ : ℚ⁺)`; this follows Definition 11.2.10 from {{#cite UF13}}.

Reviewer: ``The comment

> this follows Definition 11.2.10 from {{#cite UF13}}.

fits better in the `References` section imo

This holds if and only if any `ε : ℚ⁺` is an is an upper bound on the distance
between `f δ` and `l` for all positive rational numbers `δ < ε`.

## Definitions

### Limits of a Cauchy approximation in a premetric space

```agda
module _
  {l1 l2 : Level} (A : Premetric-Space l1 l2)
  (f : cauchy-approximation-Premetric-Space A)
  (l : type-Premetric-Space A)
  where

  is-limit-cauchy-approximation-prop-Premetric-Space : Prop l2
  is-limit-cauchy-approximation-prop-Premetric-Space =
    Π-Prop
      ( ℚ⁺)
      ( λ ε →
        Π-Prop
          ( ℚ⁺)
          ( λ δ →
            structure-Premetric-Space
              ( A)
              ( ε +ℚ⁺ δ)
              ( map-cauchy-approximation-Premetric-Space A f ε)
              ( l)))

  is-limit-cauchy-approximation-Premetric-Space : UU l2
  is-limit-cauchy-approximation-Premetric-Space =
    type-Prop is-limit-cauchy-approximation-prop-Premetric-Space

  is-prop-is-limit-cauchy-approximation-Premetric-Space :
    is-prop is-limit-cauchy-approximation-Premetric-Space
  is-prop-is-limit-cauchy-approximation-Premetric-Space =
    is-prop-type-Prop is-limit-cauchy-approximation-prop-Premetric-Space
```

> ### Local approximates of Cauchy approximations

Reviewer: I can't find a definition of what an "approximate" is anywhere. Would you care to elaborate on this concept? What's the difference between an approximate, an approximation, and a bound for instance?

Author: I knew this naming was a bit clumsy.
As noted below, `limits` and what I call `local approximates` are actually equivalent. The difference is that `limits` are quantified like `∀ (ε δ : ℚ⁺) ... P (ε +ℚ⁺ δ) ε` and `approximates` are quantified like `∀ (ε δ : ℚ⁺) (δ < ε) .. P ε δ`.

Depending on the context, one can be slightly easier to work with than the other, e.g. `is-indistinguishable-is-approximate-cauchy-approximation-triangular-symmetric-Premetric-Space` (but maybe my proof is more complicated that it should be?).

The idea that the second approach quantifies over "small δ's" and that, for these small δ's, `f δ` is near the limit value, made me chose this name. But I'm open to suggestions.

Reviewer: I see. From what I can gather, `approximate` is either an adjective or a verb, but here you seem to be using it as a noun, which would be wrong. Why do you not want to use "approximation"?

Reviewer: I see that approximation would be used in two different meanings here then

Author: `is-local-approximation-cauchy-approximation` seems a bit redundant. what about "estimate" ?

Reviewer: What about something like "is locally a limiting point" or "is locally a convergence point"?

Reviewer: "estimate" is good to me

Author: > "estimate" is good to me

yes. I think it's the word I was looking for from the beginning.
see 8f2a4e3d79755ec82a2d93516453b8c06ca4a058 for a fix.
I like the dual interpretation "`l` is a limit of `f`" / "`f δ` is an estimate of `l`".

```agda
module _
  {l1 l2 : Level} (A : Premetric-Space l1 l2)
  (f : cauchy-approximation-Premetric-Space A)
  (l : type-Premetric-Space A)
  where

  is-approximate-prop-cauchy-approximation-Premetric-Space : Prop l2
  is-approximate-prop-cauchy-approximation-Premetric-Space =
    Π-Prop
      ( ℚ⁺)
      ( λ ε →
        Π-Prop
          ( ℚ⁺)
          ( λ δ →
            hom-Prop
              ( le-prop-ℚ⁺ δ ε)
              ( structure-Premetric-Space
                ( A)
                ( ε)
                ( map-cauchy-approximation-Premetric-Space A f δ)
                ( l))))

  is-approximate-cauchy-approximation-Premetric-Space : UU l2
  is-approximate-cauchy-approximation-Premetric-Space =
    type-Prop is-approximate-prop-cauchy-approximation-Premetric-Space

  is-prop-is-approximate-cauchy-approximation-Premetric-Space :
    is-prop is-approximate-cauchy-approximation-Premetric-Space
  is-prop-is-approximate-cauchy-approximation-Premetric-Space =
    is-prop-type-Prop is-approximate-prop-cauchy-approximation-Premetric-Space
```

## Properties

> ### Equivalence of limits and local approximates of Cauchy approximations in premetric spaces

Reviewer: Do you mean to say limits and local approximates are equivalent?

```suggestion
### Limits and local approximates of Cauchy approximations are equivalent
```

    is-prop-type-Prop is-limit-cauchy-approximation-prop-Premetric-Space
```

### Estimates of Cauchy approximations

```agda
module _
  {l1 l2 : Level} (A : Premetric-Space l1 l2)
  (f : cauchy-approximation-Premetric-Space A)
  (l : type-Premetric-Space A)
  where

  is-estimate-prop-cauchy-approximation-Premetric-Space : Prop l2
  is-estimate-prop-cauchy-approximation-Premetric-Space =
    Π-Prop
      ( ℚ⁺)
      ( λ ε →
        Π-Prop
          ( ℚ⁺)
          ( λ δ →
            hom-Prop
              ( le-prop-ℚ⁺ δ ε)
              ( structure-Premetric-Space
                ( A)
                ( ε)
                ( map-cauchy-approximation-Premetric-Space A f δ)
                ( l))))

  is-estimate-cauchy-approximation-Premetric-Space : UU l2
  is-estimate-cauchy-approximation-Premetric-Space =
    type-Prop is-estimate-prop-cauchy-approximation-Premetric-Space

  is-prop-is-estimate-cauchy-approximation-Premetric-Space :
    is-prop is-estimate-cauchy-approximation-Premetric-Space
  is-prop-is-estimate-cauchy-approximation-Premetric-Space =
    is-prop-type-Prop is-estimate-prop-cauchy-approximation-Premetric-Space
```

## Properties

> ### Limits and estimates of Cauchy approximations are equivalent

Reviewer: This statement doesn't make conceptual sense. An _estimate_ should be a limit within some possibly non-zero bound. Maybe it is better to call it a _limiting estimate_?

Reviewer: Again I must reiterate that it is important we don't misuse terms that have a pre-established meaning in the literature. Clearly, an estimate in the sense given here is not an estimate in the sense it is usually understood to be.

Author: The way I saw it, the second interpretation states that 
``` 
∀ (ε : ℚ⁺) ∀ (δ : ℚ⁺) (H : le-ℚ⁺ δ ε) → dist (f δ) l ≤ ε
```
So, for small values of `δ`, `f δ` is within a `ε`-neighborhood of `l` , hence an estimate of it (or each one is an estimate of the other, if you add symmetricity).

The canonical example would we a Cauchy approximation `f` in `ℚ` that you compose with `rational-ℝ` to get a Cauchy approximation in `ℝ`. Then the values of `f δ` for small `δ`s will be rational estimates of the limit.

Reviewer: If I have, say, the sequence $f(n) = \frac{1}{n}$, then I would say that the number $2^{-10}$ is an estimate of the limit of the sequence, but $2^{-10}$ is not the limit of the sequence. It's a different concept.

Reviewer: Wait, is "being a limit" a property of the approximation `f` or a property of the point `l` as you see it?

Author: > Is "being a limit" a property of the approximation `f` or a property of the point `l` as you see it?

Can I say "both" 😅 ? I know it's usually seen as a property of the sequence/function/Cauchy-approximation but, strictly speaking, its a property on pairs `(f , l)`, isn't it? And in some context, it's also useful to fix some `l` and consider "all the sequences with limit `l`" etc.

It's a bit like with premetrics, is `B d x y` a property of `x`, `y` or `d`? Fix `x` and `d`, and you get the ball of center `x` and radius `d`, which is interesting to define topologies, continuity, etc. Fix `x` and `y` and you get the upper bound on distances, indistinguishability, etc. There's no intrinsic point of view.

> If I have, say, the sequence $f(n)=\frac{1}{n'}$ , then I would say that the number $2^{-10}$ is an estimate of the limit of the sequence, but $2^{-10}$ is not the limit of the sequence. It's a different concept.

Well, yeah $2^{-10} = f(2^{10})$ is an estimate of the limit, and so is $f(N)$ for $N > 2^{10}$ I guess, in the same way $f(\delta)$ is an estimate of the limit of a convergent Cauchy approximation, for small $\delta$ 's. (Keep in mind that Cauchy approximations converge with $r \rightarrow 0$ in $\mathbb{Q}^+$ unlike sequences that converge with $n \rightarrow +\infty$).

Author: So yeah, this should probably be read as "`l` is a limit of `f` iff `f` is an estimate of `l`, provided I've given sense to what "being an estimate" means. 

Maybe we should start with that: given a value $l$ in a metric space $M$, an _estimate_ of $l$ is a map $f~:~ \mathbb{Q}^{+} \rightarrow M$ such that ... (same definition as limit). Then with good circumstances, any _estimate_ will be a Cauchy approximation, and then we can ask ourselves "when do Cauchy approximation have a limit", etc. These are just random thoughts, and this definitely doesn't solve my naming problem, but it may be something to consider.

Reviewer: Why wouldn't you say just _$f$ approximates $l$_, or _$f$ converges at $l$_? I don't understand why you want to use a new word for this that has a different meaning.

Reviewer: Yes!

Author: see 8be1a4f83bbe4aa51f11cd8002a9fda2c9fb9333
